runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -0,0 +1,1529 @@
+	}
+	vmFuncs->internalExitVMToJNI(currentThread);
+}
+
+/**
+ * static native void setCallSiteTargetVolatile(CallSite site, MethodHandle target);
+ */
+void JNICALL
+Java_java_lang_invoke_MethodHandleNatives_setCallSiteTargetVolatile(JNIEnv *env, jclass clazz, jobject callsite, jobject target)
+{
+	J9VMThread *currentThread = (J9VMThread*)env;
+	const J9InternalVMFunctions *vmFuncs = currentThread->javaVM->internalVMFunctions;
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+
+	if ((NULL == callsite) || (NULL == target)) {
+		vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+	} else {
+		j9object_t callsiteObject = J9_JNI_UNWRAP_REFERENCE(callsite);
+		j9object_t targetObject = J9_JNI_UNWRAP_REFERENCE(target);
+
+		UDATA offset = (UDATA)vmFuncs->instanceFieldOffset(currentThread, J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, callsiteObject), (U_8*)"target", strlen("target"), (U_8*)"Ljava/lang/invoke/MethodHandle;", strlen("Ljava/lang/invoke/MethodHandle;"), NULL, NULL, 0);
+		MM_ObjectAccessBarrierAPI objectAccessBarrier = MM_ObjectAccessBarrierAPI(currentThread);
+		objectAccessBarrier.inlineMixedObjectStoreObject(currentThread, callsiteObject, offset, targetObject, true);
+	}
+	vmFuncs->internalExitVMToJNI(currentThread);
+}
+
+/**
+ * static native void copyOutBootstrapArguments(Class<?> caller, int[] indexInfo,
+												int start, int end,
+												Object[] buf, int pos,
+												boolean resolve,
+												Object ifNotAvailable);
+ */
+void JNICALL
+Java_java_lang_invoke_MethodHandleNatives_copyOutBootstrapArguments(JNIEnv *env, jclass clazz, jclass caller, jintArray indexInfo, jint start, jint end, jobjectArray buf, jint pos, jboolean resolve, jobject ifNotAvailable)
+{
+	J9VMThread *currentThread = (J9VMThread*)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	const J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+
+	if ((NULL == caller) || (NULL == indexInfo) || (NULL == buf)) {
+		vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGINTERNALERROR, NULL);
+	} else {
+		J9Class *callerClass = J9VM_J9CLASS_FROM_JCLASS(currentThread, caller);
+		j9array_t indexInfoArray = (j9array_t)J9_JNI_UNWRAP_REFERENCE(indexInfo);
+		j9array_t bufferArray = (j9array_t)J9_JNI_UNWRAP_REFERENCE(buf);
+
+		if ((J9INDEXABLEOBJECT_SIZE(currentThread, indexInfoArray) < 2)) {
+			vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGINTERNALERROR, NULL);
+		} else if (((start < -4) || (start > end) || (pos < 0)) || ((jint)J9INDEXABLEOBJECT_SIZE(currentThread, bufferArray) <= pos) || ((jint)J9INDEXABLEOBJECT_SIZE(currentThread, bufferArray) <= (pos + end - start))) {
+			vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGLINKAGEERROR, NULL);
+		} else {
+			/* U_16 bsmArgCount = (U_16)J9JAVAARRAYOFINT_LOAD(currentThread, indexInfoArray, 0); */
```

Comment:
Why the commented code?

Reply:
This should be identical to the argCount value from bsmData, I double checked against OJDK behavior and added a check that user value should be equal to jvm value.